src/gui/macOS/fileprovider_mac.mm

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2022 by Claudio Cambra <claudio.cambra@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#import <Foundation/Foundation.h>
+
+#include <QLoggingCategory>
+
+#include "fileprovider.h"
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcMacFileProvider, "nextcloud.gui.macfileprovider", QtInfoMsg)
+
+namespace Mac {
+
+static FileProvider *_instance = nullptr;
+
+FileProvider::FileProvider(QObject * const parent)
+    : QObject(parent)
+{
+    if (!fileProviderAvailable()) {
+        qCDebug(lcMacFileProvider) << "File provider system is not available on this version of macOS.";
+        return;
+    }
+
+    qCDebug(lcMacFileProvider) << "Initialising file provider domain manager.";
+    _domainManager = std::make_unique<FileProviderDomainManager>(new FileProviderDomainManager(this));
+
+    if (_domainManager) {
+        qCDebug(lcMacFileProvider()) << "Initialized file provider domain manager";
+    }
+
+    qCDebug(lcMacFileProvider) << "Initialising file provider socket server.";
+    _socketServer = std::make_unique<FileProviderSocketServer>(new FileProviderSocketServer(this));
+
+    if (_socketServer) {
+        qCDebug(lcMacFileProvider) << "Initialised file provider socket server.";
+    }
+}
+
+FileProvider *FileProvider::instance()
+{
+    if (!fileProviderAvailable()) {
+        return nullptr;
+    }
+
+    if (!_instance) {
+        _instance = new FileProvider();
```

Comment:
you should not store it there as you are putting it inside a `QScopedPointer` in some other places and would risk this one to be invalid without a way to notice it

Reply:
Okay, changed this aspect of the class to emulate FolderMan instead